### DISCUSSION TITLE
How to pass an initialization param to a capsule from build?

### DISCUSSION THREAD

**Discussion by opsb** [GregoryConrad/rearch-dart#184] [Q&A] [answered]
I've been trying to figure out how to pass parameters in when creating a capsule. Here's my two attempts:

1. Pass `use` into capsule function:

```dart
(String, Function(String)) counter(CapsuleHandle use, int initialValue) {
  return use.state(initialValue);
}
```

```dart
RearchBuilder(builder: (context, use) {
  final (count, setCount) = counter(use, 3);
});
```

This results in a compilation error because `use` is a `WidgetHandle` not a `CapsuleHandle`. Switching it to expect a `WidgetHandle` doesn't help because it doesn't have the `use.state` function available.

2. Use a factory

```dart
(String, Function(String)) Function(int) counter(CapsuleHandle use) {
  return (initialCount) => use.state(initialCount)
}
```

```dart
final (count, setCount) = use(counter)(3);
```
This time there's a runtime error `You may only register side effects during a capsule's build!`.

What's the right approach here?

**Comment by busslina** [comment]
You can't modify the capsule signature, so you can't pass initialization params to a capsule. You can read params from another capsule or you can modify a capsule (like the example `set` or `increment`). To pass initialization params seems to be imperative, so you should find a declarative way to do it.
You can pass initialization params to a custom side effect, but not to a capsule.

```dart
CounterController counterCapsule(CapsuleHandle use) {
  final count = use.data(0);

  return CounterController(
    count: count.value,
    increment: () => count.value++,
    set: (newValue) => count.value = newValue,
  );
}

class CounterController {
  CounterController({
    required this.count,
    required this.increment,
    required this.set,
  });

  final int count;
  final void Function() increment;
  final void Function(int value) set;
}

main() {
  final container = CapsuleContainer();

  final counter = container.read(counterCapsule);

  counter.set(3);
}
```

**Reply by busslina** [reply]
BTW, it's recommended to use `data` instead of `state`.

**Reply by opsb** [reply]
Interesting, what's the rationale for that?

**Reply by busslina** [reply]
> Interesting, what's the rationale for that?

If you mean about `state/data`, `state` was first introduced and not too much time ago Greg introduced `data` at the same time he modified the core to allow to modify the current capsule state (data) when building the capsule. Prior to that (state) you couldn't do that because an error came after that (recursive building). So now, with data you can do almost everything: modify current state on building or modify capsule state from outside (executing a returned controller callback). One thing to consider is that it seems better to return a getter instead of a value, because the getter will give you always the latest value (even if you cache the getter). So `use.data` returns a getter/setter while `use.state` returns a value/setter.

**Reply by busslina** [reply]
We can say `data` and `state` is almost the same., But `data` is like a replacement for `state` and it's marked as:
```
/// While [state] will continue to work just fine,
/// it is recommended that newer applications also take a look at [data].
```

**Reply by GregoryConrad** [reply]
As @busslina mentioned, the idea is that `data` is more flexible than `state`. You can update `data` within or externally to a build using the same exact API and all will continue to work just fine.

For some (pretty recent) history, before there was a `use.rawValueWrapper` that I was never happy with, and one had to manually call `rebuild` to get states to update. Now, no new code should ever touch `rebuild` directly. I _would_ remove it, but that'd break backwards compatibility. Perhaps I should deprecate it though--that's an idea.

`state` is kept because:
1. backwards compatibility reasons
2. it's exactly the api you want in certain simple scenarios, so you can just do `=> use.state(123);`
3. for folks coming from react/flutter hooks

**Reply by opsb** [reply]
Coming from react hooks it certainly seemed like `use.state` would be the preferred way to do it. From what you say deprecating it would be helpful for newcomers (possibly even permanently as it's a good breadcrumb for react folk).

**Comment by GregoryConrad** [comment, marked as answer]
You cannot pass parameters to a capsule because it would no longer be a capsule (and capsules are typedefs for `T Function(CapsuleHandle)`). For many situations, the answer is factory capsules. There are also "dynamic capsules," but unless you need incremental computation (you will know if you need it when you need it) then you don't need it. For yours, it looks like you just want another capsule:

```dart
ValueWrapper<int> startingCountManager(CapsuleHandle use) => use.data(0);

(int, void Function()) countManager(CapsuleHandle use) {
  final count = use.data(use(startingCountManager).value);
  return (count.value, () => count.value++);
}
```

And just make sure to read/set the `startingCountManager` before `countManager` so the default is used.

**Reply by opsb** [reply]
There's two scenarios where I'm looking for a solution where it isn't possible to rely on another capsule.

1. A widget has a parameter which I need to pass to a capsule. It's a large codebase so I'm looking for an incremental adoption path for Rearch. In many cases this would seem to require values being passed in from outside of Rearch's capsule graph.

2. The family/factory scenario where there are multiple instances. Factories seem like a good fit but unfortunately it isn't possible to use the factory params with `use` (the `You may only register side effects during a capsule's build!` error)

How would you approach these? Perhaps dynamic modules might help? Perhaps they could be modelled as side effects operating on the widget params as remote state?

**Reply by GregoryConrad** [reply]
> A widget has a parameter which I need to pass to a capsule.

Oh, that would normally strike me as a no-no in ReArch. If this is a parameter from a widget, I have two questions for you:
1. Should this actually be local or scoped widget state, and not a (global) capsule? This is perhaps what you actually need here (unless you're sure the capsule is for app-level state).
2. If you're sure it should be global, can you set the default value from wherever the origin of this parameter is? I.e., if a parent widget is what created the value, can you have the parent watch/update the `startingCountManager` capsule itself?

> The family/factory scenario where there are multiple instances. Factories seem like a good fit but unfortunately it isn't possible to use the factory params with use (the You may only register side effects during a capsule's build! error)

Yes, this is because individual instances of a factory cannot register side effects. You can do:
```dart
int Function() myIntFactory(CapsuleHandle use) {
  use.myFancySideEffect();
  return () => 123;
}
```
But `use` must always be before any asynchronous gap (start of the callback function above), and side effects in factories should also be very rare.

If you need multiple _stateful instances_, you actually either need:
1. multiple global capsules (in case the state is actually global for some reason, which is normally highly unlikely). An example of this is the pickleball example application, where we have two teams that each need their own state
2. scoped state to some section of the widget tree (normally the correct solution)

**Reply by GregoryConrad** [reply]
@opsb based on what you've described thus far, I'm guessing you probably want scoped state:

```dart
@rearchInheritedWidget
(int, void Function(int)) scopedCount(WidgetHandle use, int startingCount) => use.state(startingCount);

MyFancyParentWidget(
  ScopedCount(123, child: MyFancyChildWidget(...)),
);
```

**Reply by opsb** [reply]
aha, yes I believe the Scoping State docs where I get the idea for version (1) in my OP. I do have one case where the InheritedWidget approach makes a lot of sense so that's very helpful. Based on what I've seen in the docs there would it not also be possible to do the following? (very similar to my first approach above, just with the WidgetHandle instead of a CapsuleHandle).

```dart
(String, Function(String)) counter(WidgetHandle use, int initialValue) {
  return use.state(initialValue);
}
```
```dart
RearchBuilder(builder: (context, use) {
  final (count, setCount) = counter(use, 3);
});
```

**Reply by GregoryConrad** [reply]
You can do that, but it’s more idiomatic to make an `extension on SideEffectRegistrar` instead of making a method that consumes a WidgetHandle there

**Reply by opsb** [reply]
I suppose the way to think about this is not creating a parameterised capsule, rather just having a helper which uses an existing side effect.

**Reply by opsb** [reply]
Right, gotcha, I haven't used that mechanism yet, makes sense

**Reply by GregoryConrad** [reply]
> just having a helper which uses an existing side effect

Oh sure, that’s just making a new side effect, which is done with extension methods on SideEffectRegisrar/WidgetSideEffectRegistrar

**Reply by opsb** [reply]
Awesome, thanks!
